earthly/docs/Earthfile

```diff
@@ -0,0 +1,175 @@
+
+    # Copy our dependencies.
+    COPY pyproject.toml .
+    COPY poetry.lock .
+    COPY start.sh .
+
+    # Install it all with poetry
+    RUN poetry install --only main --no-root
+
+    # Trust directory, required for git >= 2.35.2
+    RUN git config --global --add safe.directory /docs &&\
+        git config --global --add safe.directory /site
+
+    # Carry the Kroki Docker Compose with this image.
+    WORKDIR /kroki
+    COPY docker-compose.kroki.yml ./docker-compose.yml
+
+    # Set working directory
+    WORKDIR /docs
+
+    # Expose MkDocs development server port
+    EXPOSE 8000
+
+    # Start development server by default
+    ENTRYPOINT ["/poetry/start.sh"]
+    CMD ["serve", "--dev-addr=0.0.0.0:8000"]
+
+    # Finish building the container, and publish to the registry
+    IF [ "$registry" = "" ]
+        ARG registry_final=$registry
+    ELSE
+        ARG registry_final=${registry}/
+    END
+
+    SAVE ARTIFACT /kroki/docker-compose.yml /kroki/docker-compose.yml
+
+    SAVE IMAGE --push ${registry_final}mkdocs-material:${tag}
+
+MKDOCS_BUILD:
+    COMMAND
+    ARG src
+    ARG dest
+
+    # Ensure the mkdocs tools are built and up-to-date.
+    BUILD +mkdocs-material
+
+    FROM earthly/dind:alpine
+    WORKDIR /build
+    COPY +mkdocs-material/kroki/docker-compose.yml .
+
+    # Copy the src
+    COPY --dir $src /docs
+
+    # Docs need Kroki to build diagrams, so make sure its running.
+    WITH DOCKER --compose docker-compose.yml --load mkdocs=+mkdocs-material
+        RUN docker run --rm -v /docs:/docs --network=default_mkdocs mkdocs build
+    END
+
+    SAVE ARTIFACT /docs/site AS LOCAL $dest
+
+MKDOCS_SERVE:
+    COMMAND
+    ARG src
+    ARG port
+
+    # Ensure the mkdocs tools are built and up-to-date.
+    BUILD +mkdocs-material
+
+    FROM earthly/dind:alpine
+    WORKDIR /build
+    COPY +mkdocs-material/kroki/docker-compose.yml .
+
+    # Copy the src
+    COPY --dir $src /docs
+
+    # Docs need Kroki to build diagrams, so make sure its running.
+    WITH DOCKER --compose docker-compose.yml --load mkdocs=+mkdocs-material
+        RUN --privileged docker run --rm -p $port:8000 -v /docs:/docs --network=default_mkdocs mkdocs
+    END
+
+MDLINT:
+    # Linting is done with MarkdownLint CLI2
+    # See: https://github.com/DavidAnson/markdownlint-cli2
+    COMMAND
+
+    ARG fix
+    ARG src
+
+    RUN --no-cache echo SRC: $src
+
+    # cspell: words davidanson
+    WITH DOCKER --pull davidanson/markdownlint-cli2-rules:next
+        #RUN --no-cache docker run --rm -v $src:/workdir --entrypoint sh davidanson/markdownlint-cli2-rules:next -c "ls -al; ls -al docs"
+        RUN --no-cache docker run --rm -v $src:/workdir davidanson/markdownlint-cli2-rules:next "**/*.md" --config .markdownlint-cli2.jsonc $fix
+    END
+
+CSPELL:
+    # Spell checking all code is done with cspell
+    # See: cspell.org
+    COMMAND
+
+    ARG src
+
+    WITH DOCKER --pull ghcr.io/streetsidesoftware/cspell:latest
```

Comment:
I'm a little torn on this approach. It seems DinD can be a bit buggy with Earthly sometimes, and troubleshooting it is not trivial. Are you opposed to just pulling these tools in using their natural patterns instead of running their containers? It's a bit more work, but then you can easily `earthly -i +target` and get into  sane environment to test an issue rather than having to navigate through DinD.

Reply:
Its a bit of work to get it right, and then we have to maintain it.
If we think its necessary, I would rather that be the subject of an alternative PR.
For now, I have moved both linters to their own earthfiles, so this kind of refactoring is easier.
I would advocate we leave it for now, as I have bigger fish to fry, and we revisit it if it causes breakage in CI.